File: 2025-04-April-GFG-POTD/April 05 - Topological sort/Solution.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    vector<int> topoSort(int V, vector<vector<int>>& edges) {
        vector<int> result;
        vector<vector<int>> adj(V);
        
        // Build adjacency list
        for (auto edge : edges) {
            adj[edge[0]].push_back(edge[1]);
        }
        
        // Calculate in-degrees
        vector<int> inDegree(V, 0);
        for (int u = 0; u < V; ++u) {
            for (int v : adj[u]) {
                inDegree[v]++;
            }
        }
        
        // Queue for nodes with 0 in-degree
        queue<int> q;
        for (int i = 0; i < V; ++i) {
            if (inDegree[i] == 0) q.push(i);
        }
        
        while (!q.empty()) {
            int node = q.front(); q.pop();
            result.push_back(node);
            for (int neighbor : adj[node]) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) q.push(neighbor);
            }
        }
        
        return result;
    }
};

// Usage Example
int main() {
    int V = 4, E = 3;
    vector<vector<int>> edges = {{3, 0}, {1, 0}, {2, 0}};
    Solution obj;
    vector<int> result = obj.topoSort(V, edges);

    for (int node : result) {
        cout << node << " ";
    }
    cout << endl;
    return 0;
}

File: 2025-04-April-GFG-POTD/April 05 - Topological sort/Solution.py
from collections import deque

class Solution:
    def topoSort(self, V, edges):
        result = []
        adj = [[] for _ in range(V)]

        # Build adjacency list
        for u, v in edges:
            adj[u].append(v)

        # Calculate in-degrees
        in_degree = [0] * V
        for u in range(V):
            for v in adj[u]:
                in_degree[v] += 1

        # Queue for nodes with 0 in-degree
        q = deque()
        for i in range(V):
            if in_degree[i] == 0:
                q.append(i)

        while q:
            node = q.popleft()
            result.append(node)
            for neighbor in adj[node]:
                in_degree[neighbor] -= 1
                if in_degree[neighbor] == 0:
                    q.append(neighbor)

        return result

# Usage Example
if __name__ == "__main__":
    V = 4
    E = 3
    edges = [[3, 0], [1, 0], [2, 0]]
    sol = Solution()
    res = sol.topoSort(V, edges)
    print(res)
